[3.x] Add template annotations

WyriHaximus opened the pull request:
Adds template annotations turning the `PromiseInterface` into a generic.

Variables `$p1` and `$p2` in the following code example both are `PromiseInterface<int|string>`.

```php
$f = function (): int|string {
    return time() % 2 ? 'string' : time();
};

/**
 * @return PromiseInterface<int|string>
 */
$fp = function (): PromiseInterface {
    return resolve(time() % 2 ? 'string' : time());
};

$p1 = resolve($f());
$p2 = $fp();
```

When calling `then` on `$p1` or `$p2`, PHPStan understand that function `$f1` is type hinting its parameter fine, but `$f2` will throw during runtime:

```php
$p2->then(static function (int|string $a) {});
$p2->then(static function (bool $a) {});
```

Builds on top of https://github.com/reactphp/promise/pull/246 and https://github.com/reactphp/promise/pull/188 and is a requirement for https://github.com/reactphp/async/pull/40

clue commented on .github/workflows/ci.yml:
```diff
@@ -52,3 +52,6 @@ jobs:
       - run: composer install
       - run: vendor/bin/phpstan
+        if: ${{ matrix.php >= 7.2 }}
+      - run: vendor/bin/phpstan --configuration="phpstan.legacy.neon.dist"
```
LGTM :+1: 

Note that this file needs to be added to `.gitattributes` (#153).

clue commented on phpstan.legacy.neon.dist:
```diff
@@ -0,0 +1,7 @@
+parameters:
+	ignoreErrors:
+		- '#Template type T is declared as covariant, but occurs in contravariant position in parameter result of method React\\Promise\\Promise::settle\(\).#'
+		- '#Template type T is declared as covariant, but occurs in contravariant position in parameter promise of method React\\Promise\\Promise::unwrap\(\).#'
```
LGTM, but we may want to double-check this is still needed with all changes suggested below :+1:

WyriHaximus replied:
No need to double check, PHPStan will error on this when those errors aren't showing up anymore.

clue commented on phpstan.neon.dist:
```diff
@@ -4,3 +4,4 @@ parameters:
     paths:
         - src/
         - tests/
+        - types/
```
Note that this directory needs to be added to `.gitattributes` (#153) if we keep this.

We've also discussed, it probably makes sense to move this to `tests/` instead. We may also want to break this down into individual files like `FunctionResolveTypes.php` etc.

WyriHaximus replied:
Moved it into the tests folder, and will also have a look at breaking into different files

clue commented on src/Deferred.php:
```diff
@@ -21,6 +26,9 @@ public function __construct(callable $canceller = null)
     }
 
+    /**
+     * @return PromiseInterface<T>
```
It's my understanding the `resolve()` method defines the type here, no?

Should probably add some tests similar to this:

```php
$deferred = new Deferred();
assertType('React\Promise\PromiseInterface<T>', $deferred->promise());
```

```php
$deferred = new Deferred();
$deferred->resolve(42);
assertType('React\Promise\PromiseInterface<int>', $deferred->promise());
```

Not sure how much of this is supported, WDYT?

WyriHaximus replied:
Added it, it's picking both up as `mixed`.

clue commented on src/Internal/RejectedPromise.php:
```diff
@@ -8,6 +8,9 @@
  * @internal
+ *
+ * @template-implements PromiseInterface<T>
+ * @template-covariant T
```
It's my understanding this should read like `RejectedPromise<never> implements PromiseInterface<never>`, no?

Should probably add some tests similar to this:

```php
assertType('React\Promise\PromiseInterface<never>', new RejectedPromise(new \RuntimeException()));
```

Not sure how much of this is supported, WDYT?

WyriHaximus replied:
Not sure if this is a yes or no:

```
 ------ ------------------------------------------------------------------------------------------------------ 
  Line   tests/types/reject.php                                                                                
 ------ ------------------------------------------------------------------------------------------------------ 
  6      Expected type React\Promise\PromiseInterface<never>, actual: React\Promise\PromiseInterface<*NEVER*>  
 ------ ------------------------------------------------------------------------------------------------------ 
```

clue commented on src/Promise.php:
```diff
@@ -4,12 +4,16 @@
     private $canceller;
 
-    /** @var ?PromiseInterface */
+    /** @var PromiseInterface<T> */
```
This needs to be nullable for pending promises.

clue commented on src/Promise.php:
```diff
@@ -77,11 +81,11 @@ public function catch(callable $onRejected): PromiseInterface
     {
         return $this->then(static function ($value) use ($onFulfilledOrRejected) {
-            return resolve($onFulfilledOrRejected())->then(function () use ($value) {
+            return resolve($onFulfilledOrRejected())->then(function ($_) use ($value) {
```
Is this still needed?

WyriHaximus replied:
No.

clue commented on src/Promise.php:
```diff
@@ -77,11 +81,11 @@ public function catch(callable $onRejected): PromiseInterface
             });
         }, static function ($reason) use ($onFulfilledOrRejected) {
-            return resolve($onFulfilledOrRejected())->then(function () use ($reason) {
+            return resolve($onFulfilledOrRejected())->then(function ($_) use ($reason) {
```
Is this still needed?

WyriHaximus replied:
No.

clue commented on src/Promise.php:
```diff
@@ -166,6 +170,9 @@ private function reject(\Throwable $reason): void
     }
 
+    /**
+     * @param PromiseInterface<T>|PromiseInterface<null> $result
```
The `PromiseInterface<null>` seems weird here. It looks like this should be `PromiseInterface<never>` for rejected promises? In either case, a short comment would be useful here :+1:

clue commented on src/Promise.php:
```diff
@@ -193,13 +200,17 @@ private function settle(PromiseInterface $result): void
     }
 
+    /**
+     * @param PromiseInterface<T>|PromiseInterface<null> $promise
```
Same as above, the `null` appears to be unexpected.

clue commented on src/PromiseInterface.php:
```diff
@@ -28,9 +31,9 @@ interface PromiseInterface
-     * @return PromiseInterface
+     * @template TFulfilled as PromiseInterface<T>|T
+     * @param (callable(T): TFulfilled)|null $onFulfilled
+     * @return PromiseInterface<T>
```
Shouldn't this be `@return PromiseInterface<TFulfilled>` (plus additional type checks if the function returns `PromiseInterface<TFulfilled>`).

Should probably add some tests similar to this:

```php
assertType('React\Promise\PromiseInterface<bool>', resolve(true));

assertType('React\Promise\PromiseInterface<int>', resolve(true)->then(fn(bool $value): int => (int) $value));

assertType('React\Promise\PromiseInterface<int>', resolve(true)->then(fn(bool $value): PromiseInterface => resolve((int) $value)));
```

Not sure how much of this is supported, WDYT?

WyriHaximus replied:
This would be really good to have, it is also a really hard one. Here is the PHPStan playground with my work on it so far: https://phpstan.org/r/bf62f95b-4560-4115-9eac-15518640f137

Shifting my focus to your other comments for the rest of today before coming back to this one.

clue replied:
Resolved :shipit:

clue commented on src/PromiseInterface.php:
```diff
@@ -44,8 +47,7 @@ public function then(?callable $onFulfilled = null, ?callable $onRejected = null
      *
-     * @param callable $onRejected
-     * @return PromiseInterface
+     * @return PromiseInterface<T>
```
Shouldn't this be `@return PromiseInterface<TFulfilled>`? (See `then()` above)

WyriHaximus replied:
Will make sure those two align

clue commented on src/PromiseInterface.php:
```diff
@@ -91,8 +93,9 @@ public function catch(callable $onRejected): PromiseInterface;
-     * @param callable $onFulfilledOrRejected
-     * @return PromiseInterface
+     * @template TReturn of PromiseInterface<T>|T
+     * @param callable(T=): TReturn $onFulfilledOrRejected
```
The `$onFulfilledOrRejected` doesn't receive any arguments and its return value is ignored (unless it throws).

clue commented on src/PromiseInterface.php:
```diff
@@ -118,7 +121,7 @@ public function cancel(): void;
      *
      * @param callable $onRejected
-     * @return PromiseInterface
+     * @return PromiseInterface<T>
```
See `catch()` above.

clue commented on src/PromiseInterface.php:
```diff
@@ -135,7 +138,7 @@ public function otherwise(callable $onRejected): PromiseInterface;
      *
      * @param callable $onFulfilledOrRejected
-     * @return PromiseInterface
+     * @return PromiseInterface<T>
```
See `finally()` above.

clue commented on src/functions.php:
```diff
@@ -54,8 +56,7 @@ function resolve($promiseOrValue): PromiseInterface
  *
- * @param \Throwable $reason
- * @return PromiseInterface
+ * @return PromiseInterface<null>
```
Shouldn't this be `PromiseInterface<never>`?

Should probably add some tests similar to this:

```php
assertType('React\Promise\PromiseInterface<never>', reject(new \RuntimeException()));

assertType('React\Promise\PromiseInterface<never>', reject(new \RuntimeException())->then(fn(): int => 42));

assertType('React\Promise\PromiseInterface<int>', reject(new \RuntimeException())->then(null, fn(): int => 42));

assertType('React\Promise\PromiseInterface<int>', reject(new \RuntimeException())->catch(fn(): int => 42));
```

Not sure how much of this is supported, WDYT?

WyriHaximus replied:
Same issue as fullfilled promise type chaging during the chain:

```
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   tests/types/reject.php                                                                                                                                                                      
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  8      Expected type React\Promise\PromiseInterface<never>, actual: React\Promise\PromiseInterface<*NEVER*>                                                                                        
  9      Expected type React\Promise\PromiseInterface<never>, actual: React\Promise\PromiseInterface<*NEVER*>                                                                                        
  9      Parameter #1 $onFulfilled of method React\Promise\PromiseInterface<*NEVER*>::then() expects (callable(never): (React\Promise\PromiseInterface<never>|void))|null, 'React\\Promise\\Types…'  
         given.                                                                                                                                                                                      
  10     Expected type React\Promise\PromiseInterface<int>, actual: React\Promise\PromiseInterface<*NEVER*>                                                                                          
  11     Expected type React\Promise\PromiseInterface<int>, actual: React\Promise\PromiseInterface<*NEVER*>                                                                                          
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

```

WyriHaximus replied:
So the trick for line 8 and 9 is to literally use `*NEVER*` instead of `never` :facepalm:

clue commented on src/functions.php:
```diff
@@ -87,7 +89,7 @@ function all(iterable $promisesOrValues): PromiseInterface
 
             resolve($promiseOrValue)->then(
-                function ($value) use ($i, &$values, &$toResolve, &$continue, $resolve): void {
+                function ($value) use ($i, &$values, &$toResolve, &$continue, $resolve): void { /** @phpstan-ignore-line */
```
Is this still needed?

clue replied:
Resolved :shipit:

clue commented on src/functions.php:
```diff
@@ -132,7 +135,7 @@ function race(iterable $promisesOrValues): PromiseInterface
             $cancellationQueue->enqueue($promiseOrValue);
 
-            resolve($promiseOrValue)->then($resolve, $reject)->finally(function () use (&$continue): void {
+            resolve($promiseOrValue)->then($resolve, $reject)->finally(function () use (&$continue): void { /** @phpstan-ignore-line */
```
Is this still needed?

clue commented on tests/FunctionAllTest.php:
```diff
@@ -9,9 +9,9 @@ class FunctionAllTest extends TestCase
-            ->expects(self::once())
+        /** @var callable $mock */
+        $mock = $_ = $this->createCallableMock();
+        $_->expects(self::once())
```
Is this still needed? (see also below)

See alternative solution in https://github.com/reactphp/promise/pull/246/files#diff-abf1fa77a19f720052b75bc802df0260157b9064d8c3c12a80794f01e9acc065R49.

WyriHaximus replied:
Doesn't seem needed anymore, so no need for the alternative solution (currently).

clue commented on tests/Internal/CancellationQueueTest.php:
```diff
@@ -96,7 +96,7 @@ public function rethrowsExceptionsThrownFromCancel(): void
     }
 
-    private function getCancellableDeferred(): Deferred
+    private function getCancellableDeferred(): Deferred /** @phpstan-ignore-line */
```
Is this still needed? Should probably include `@return Deferred<never>`.

WyriHaximus replied:
Included that fixes it.

clue commented on tests/Internal/RejectedPromiseTest.php:
```diff
@@ -9,14 +9,17 @@
     public function getPromiseTestAdapter(callable $canceller = null): CallbackPromiseAdapter
     {
-        /** @var ?RejectedPromise */
+        /** @var ?RejectedPromise<T> */
```
Should probably be `?RejectedPromise<never>`, no?

WyriHaximus replied:
Correct.

clue commented on tests/PromiseAdapter/CallbackPromiseAdapter.php:
```diff
@@ -17,6 +17,9 @@ public function __construct(array $callbacks)
     }
 
+    /**
+     * @phpstan-ignore-next-line
```
Is this still needed? Should probably be `@return PromiseInterface<T>`, no?

WyriHaximus replied:
Yes, but see comment on `PromiseFulfilledTestTrait`. Would like to have it in, but IMHO not worth it right now.

clue replied:
Resolved :shipit:

clue commented on tests/PromiseAdapter/PromiseAdapterInterface.php:
```diff
@@ -6,6 +6,9 @@
 interface PromiseAdapterInterface
 {
+    /**
+     * @phpstan-ignore-next-line
```
Is this still needed? Should probably be `@return PromiseInterface<T>`, no?

WyriHaximus replied:
Yes, but see comment on `PromiseFulfilledTestTrait`. Would like to have it in, but IMHO not worth it right now.

clue commented on tests/PromiseTest/PromiseFulfilledTestTrait.php:
```diff
@@ -191,7 +191,7 @@ public function thenShouldContinueToExecuteCallbacksWhenPriorCallbackSuspendsFib
 
         $fiber = new \Fiber(function () use ($adapter) {
-            $adapter->promise()->then(function (int $value) {
+            $adapter->promise()->then(function (int $value) { /** @phpstan-ignore-line */
```
Is this still needed?

WyriHaximus replied:
Removing that yields the following error:

```
Parameter #1 $onFulfilled of method React\Promise\PromiseInterface<mixed>::then() expects (callable(mixed): void)|null, Closure(int): void given.  
         💡 Type int of parameter #1 $value of passed callable needs to be same or wider than parameter type mixed of accepting callable. 
```

Can try and find a way to make it work, but this is one of those where I rather be lazy and ignore it to focus on the more important parts of the PR.

clue replied:
Resolved :shipit:

clue commented on types/Promises.php:
```diff
@@ -0,0 +1,84 @@
+    return resolve(time() % 2 ? 'string' : time());
+};
+
+$tosseable = new Exception('Oops I did it again!');
```
Leftover?

WyriHaximus replied:
Yes

clue reviewed (changes requested):
@WyriHaximus Solid progress! :+1: 

For context: We've had a call together with @WyriHaximus and @SimonFrings earlier today, discussing our progress for template annotations and unhandled rejections. We've discussed most of these points internally already, but figured it makes sense to post here as well for transparency.

WyriHaximus commented on tests/PromiseTest/PromiseFulfilledTestTrait.php:
```diff
@@ -261,7 +261,7 @@ public function finallyShouldNotSuppressValueWhenHandlerReturnsANonPromiseForFul
         $adapter->resolve($value);
         $adapter->promise()
-            ->finally(function () {
+            ->finally(function () { /** @phpstan-ignore-line  */
```
These, and all the other ignores on finally method calls are there because it's part of the test to break the expected contract.

WyriHaximus committed:
[3.x] Add template annotations

Adds template annotations turning the `PromiseInterface` into a generic.

Variables `$p1` and `$p2` in the following code example both are
`PromiseInterface<int|string>`.

```php
$f = function (): int|string {
    return time() % 2 ? 'string' : time();
};

/**
 * @return PromiseInterface<int|string>
 */
$fp = function (): PromiseInterface {
    return resolve(time() % 2 ? 'string' : time());
};

$p1 = resolve($f());
$p2 = $fp();
```

When calling `then` on `$p1` or `$p2`, PHPStan understand that function
`$f1` is type hinting its parameter fine, but `$f2` will throw during
runtime:

```php
$p2->then(static function (int|string $a) {});
$p2->then(static function (bool $a) {});
```

Builds on top of https://github.com/reactphp/promise/pull/246 and
https://github.com/reactphp/promise/pull/188 and is a requirement for
https://github.com/reactphp/async/pull/40

clue commented on .github/workflows/ci.yml:
```diff
@@ -43,7 +43,6 @@ jobs:
           - 7.4
           - 7.3
           - 7.2
-          - 7.1
```
For the reference: Can't run PHPStan on legacy PHP 7.1 (unlike #246) because we (need to) use conditional return types which require PHPStan 1.6 which in turn requires PHP 7.2+ (https://phpstan.org/blog/phpstan-1-6-0-with-conditional-return-types)

clue commented on tests/types/resolve.php:
```diff
@@ -0,0 +1,87 @@
+// }));
+// assertType('React\Promise\PromiseInterface<*NEVER*>', resolve(true)->finally(function (): PromiseInterface {
+//     return reject(new \RuntimeException());
+// }));
```
For the reference: I'll leave this as an exercise for a future follow-up. I'm not too concerned about this at the moment, but it looks indeed like PHPStan currently only has limited support for checking the `never` return type in conditional expressions. I believe this to be forward compatible, so PRs are very much welcome! :shipit:

clue reviewed (approved):
@WyriHaximus Went over your changes together with @SimonFrings and addressed all outstanding issues, great work, now let's get this shipped! :shipit: :shipit: :shipit:

WyriHaximus commented:
@clue @SimonFrings Thank you for picking up and finishing this off and having my back in these challenging personal times :heart:  .

clue commented:
@WyriHaximus No problem, we're here! Happy to support you during tough times. Take care and stay strong!
